xarray/indexes/interval_index.py

```diff
@@ -0,0 +1,288 @@
+            if any(idx._bounds_dim != bounds_dim for idx in indexes):
+                raise ValueError(
+                    f"Cannot concatenate along dimension {dim!r} indexes with different "
+                    "boundary coordinate or dimension names"
+                )
+        else:
+            bounds_dim = "bounds"
+
+        return cls(new_mid_index, new_bounds_index, str(bounds_dim))
+
+    @property
+    def mid_index(self) -> PandasIndex:
+        return self._mid_index
+
+    @property
+    def bounds_index(self) -> PandasIndex:
+        return self._bounds_index
+
+    @property
+    def dim(self) -> Hashable:
+        return self.mid_index.dim
+
+    @property
+    def bounds_dim(self) -> Hashable:
+        return self._bounds_dim
+
+    def create_variables(
+        self, variables: Mapping[Any, Variable] | None = None
+    ) -> dict[Any, Variable]:
+        new_variables = self.mid_index.create_variables(variables)
+
+        # boundary variable (we cannot just defer to self.bounds_index.create_variables())
+        bounds_pd_index = cast(pd.IntervalIndex, self.bounds_index.index)
+        bounds_varname = bounds_pd_index.name
+        attrs: Mapping[Hashable, Any] | None
+        encoding: Mapping[Hashable, Any] | None
+
+        if variables is not None and bounds_varname in variables:
+            var = variables[bounds_varname]
+            attrs = var.attrs
+            encoding = var.encoding
+        else:
+            attrs = None
+            encoding = None
+
+        # TODO: wrap index data in a PandasIndexingAdapter subclass instead
+        # TODO: do we want to preserve the original dimension order for the boundary coordinate?
+        # (using CF-compliant order below)
+        data = np.stack([bounds_pd_index.left, bounds_pd_index.right], axis=-1)
+        new_variables[bounds_varname] = Variable(
+            (self.dim, self.bounds_dim), data, attrs=attrs, encoding=encoding
+        )
+
+        return new_variables
+
+    def should_add_coord_to_array(
+        self,
+        name: Hashable,
+        var: Variable,
+        dims: set[Hashable],
+    ) -> bool:
+        # add both the central and boundary coordinates if the dimension
+        # that they both share is present in the array dimensions
+        return self.dim in dims
+
+    def equals(self, other: Index) -> bool:
+        if not isinstance(other, IntervalIndex):
+            return False
+
+        return self.mid_index.equals(other.mid_index) and self.bounds_index.equals(
+            other.bounds_index
+        )
+
+    def join(self, other: Self, how: str = "inner") -> Self:
+        joined_mid_index = self.mid_index.join(other.mid_index, how=how)
+        joined_bounds_index = self.bounds_index.join(other.bounds_index, how=how)
+
+        assert isinstance(joined_bounds_index, pd.IntervalIndex)
+        check_mid_in_interval(
+            joined_mid_index.index, cast(pd.IntervalIndex, joined_bounds_index.index)
+        )
```

Comment:
I'm not sure that this check alone is enough to ensure we always yield meaningful results here, but combined with the check done in `IntervalIndex.reindex_like` (i.e., indexers are equal for both mid and bounds indexes) I think it should be enough?